src/workspace-data/data-table/entity-service/AttributeInput.ts

```diff
@@ -0,0 +1,307 @@
+import { TooltipTrigger } from '@terra-ui-packages/components';
+import _ from 'lodash/fp';
+import { Fragment, useEffect, useRef, useState } from 'react';
+import { div, fieldset, h, label, legend, span } from 'react-hyperscript-helpers';
+import { IdContainer, LabeledCheckbox, Link, RadioButton, Select } from 'src/components/common';
+import { icon } from 'src/components/icons';
+import * as Utils from 'src/libs/utils';
+
+import { renderInputForAttributeType } from '../shared/AttributeInput';
+import { convertAttributeValue, getAttributeType } from './attribute-utils';
+
+interface AttributeTypeInputProps {
+  label?: string;
+  value: TypeOption;
+  onChange: (newValue: TypeOption) => void;
+  entityTypes?: string[];
+  defaultReferenceEntityType?: string;
+  showJsonTypeOption?: boolean;
+}
+
+interface BaseTypeOption {
+  type: string;
+}
+
+interface ReferenceTypeOption extends BaseTypeOption {
+  entityType: string;
+}
+
+function isReferenceTypeOption(typeOption: TypeOption): typeOption is ReferenceTypeOption {
+  return 'entityType' in typeOption;
+}
+
+interface JsonTypeOption extends BaseTypeOption {
+  label: string;
+}
+
+function isJsonTypeOption(typeOption: TypeOption): typeOption is JsonTypeOption {
+  return 'label' in typeOption;
+}
+
+type TypeOption = BaseTypeOption | ReferenceTypeOption | JsonTypeOption;
+
+export const AttributeTypeInput = (props: AttributeTypeInputProps) => {
+  const type = props.value.type;
+  const labelText = props.label !== undefined && !!props.label ? props.label : 'Type';
+  const entityTypes = props.entityTypes !== undefined && !!props.entityTypes ? props.entityTypes : [];
+  const defaultReferenceEntityType =
+    props.defaultReferenceEntityType !== undefined && !!props.defaultReferenceEntityType
+      ? props.defaultReferenceEntityType
+      : null;
+
+  const showJsonTypeOption = props.showJsonTypeOption ?? false;
+
+  const typeOptions: TypeOption[] = [
+    { type: 'string' },
+    { type: 'reference', entityType: defaultReferenceEntityType ?? entityTypes[0] },
+    { type: 'number' },
+    { type: 'boolean' },
+  ];
+
+  if (showJsonTypeOption) {
+    typeOptions.push({ type: 'json', label: 'JSON' });
+  }
+
+  const sortedEntityTypes: string[] = _.sortBy(_.identity, entityTypes);
+
+  return div({ style: { marginBottom: '1rem' } }, [
+    fieldset({ style: { border: 'none', margin: 0, padding: 0 } }, [
+      legend({ style: { marginBottom: '0.5rem' } }, [`${labelText}:`]),
+      div(
+        {
+          style: {
+            display: 'flex',
+            flexFlow: 'row',
+            justifyContent: 'space-between',
+            marginBottom: '0.5rem',
+          },
+        },
+        _.map(
+          (typeOption: TypeOption) =>
```

Comment:
(I'm using this diff to compare the new .ts with the old .js: https://www.diffchecker.com/ObXPikjI/)

it looks like there are some slight logic changes here - were those needed to make typescript happy? or are they needed for something else?

Reply:
Yeah I was mostly just trying to make typescript happy and I'm hoping I didn't make any significant logic changes, or if I did, it was more logical.  Is there something in particular you noticed that changed?

Reply:
no, mostly just trying to understand what sort of things typescript would be mad about!